tests/jsunit/workspace_undo_redo_test.js

```diff
@@ -0,0 +1,386 @@
+/**
+ * @license
+ * Visual Blocks Editor
+ *
+ * Copyright 2017 Google Inc.
+ * https://developers.google.com/blockly/
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+ /**
+ * @fileoverview Tests for Blockly.Workspace.undo.
+ * @author marisaleung@google.com (Marisa Leung)
+ */
+'use strict';
+
+goog.require('goog.events.EventHandler');
+goog.require('goog.testing');
+goog.require('goog.testing.events');
+goog.require('goog.testing.MockControl');
+
+var workspace;
+var mockControl_;
+var saved_msg = Blockly.Msg.DELETE_VARIABLE;
+var saved_fire_func = Blockly.Events.fire;
+Blockly.defineBlocksWithJsonArray([{
+  "type": "get_var_block",
+  "message0": "%1",
+  "args0": [
+    {
+      "type": "field_variable",
+      "name": "VAR",
+    }
+  ]
+}]);
+
+function temporary_fireEvent(event) {
+  if (!Blockly.Events.isEnabled()) {
+    return;
+  }
+  Blockly.Events.FIRE_QUEUE_.push(event);
+  Blockly.Events.fireNow_();
+}
+
+function undoRedoTest_setUp() {
+  workspace = new Blockly.Workspace();
+  mockControl_ = new goog.testing.MockControl();
+  Blockly.Events.fire = temporary_fireEvent;
+}
+
+function undoRedoTest_setUpWithMockBlocks() {
+  undoRedoTest_setUp();
+  // Need to define this because field_variable's dropdownCreate() calls replace
+  // on undefined value, Blockly.Msg.DELETE_VARIABLE. To fix this, define
+  // Blockly.Msg.DELETE_VARIABLE as %1 so the replace function finds the %1 it
+  // expects.
+  Blockly.Msg.DELETE_VARIABLE = '%1';
+}
+
+function undoRedoTest_tearDown() {
+  mockControl_.$tearDown();
+  workspace.dispose();
+  Blockly.Events.fire = saved_fire_func;
+}
+
+function undoRedoTest_tearDownWithMockBlocks() {
+  undoRedoTest_tearDown();
+  Blockly.Msg.DELETE_VARIABLE = saved_msg;
+}
+
+/**
+ * Create a test get_var_block.
+ * @param {string} variable_name The string to put into the variable field.
+ * @return {!Blockly.Block} The created block.
+ */
+function createMockBlock(variable_name) {
+  var block = new Blockly.Block(workspace, 'get_var_block');
+  block.inputList[0].fieldRow[0].setValue(variable_name);
+  return block;
+}
+
+/**
+ * Check if a variable with the given values exists.
+ * @param {string} name The expected name of the variable.
+ * @param {string} type The expected type of the variable.
+ * @param {string} id The expected id of the variable.
+ */
+function undoRedoTest_checkVariableValues(name, type, id) {
```

Comment:
You've defined a similar function in workspace_test.js, xml_test.js, and variable_map_test.js.  You should combine them into a single function in tests/test_utilities.js.

In the interest of keeping this PR clean, leave this function here for now.  Then make a separate PR that just moves it to test_utilities and removes duplicates from your other test files.